src/lang/processor.rs

```diff
@@ -0,0 +1,1181 @@
+                        expected: "Wildcard or Literal".to_string(),
+                        found: format!("{:?}", arg),
+                        item: format!(
+                            "argument {} of custom predicate call '{}'",
+                            idx + 1,
+                            stmt_name_str
+                        ),
+                        span: Some(stmt_span),
+                    });
+                }
+            }
+        }
+    }
+
+    let mut stb = StatementTmplBuilder::new(pred.clone());
+    for arg in args {
+        stb = stb.arg(arg);
+    }
+    Ok(stb.desugar())
+}
+
+fn process_and_add_custom_predicate_to_batch(
+    pred_def_pair: &Pair<Rule>,
+    processing_ctx: &ProcessingContext,
+    cpb_builder: &mut CustomPredicateBatchBuilder,
+) -> Result<(), ProcessorError> {
+    let mut inner_pairs = pred_def_pair.clone().into_inner();
+    let name_pair = inner_pairs
+        .find(|p| p.as_rule() == Rule::identifier)
+        .unwrap();
+    let name = name_pair.as_str().to_string();
+
+    let arg_section_pair = inner_pairs
+        .find(|p| p.as_rule() == Rule::arg_section)
+        .unwrap();
+
+    let mut public_arg_strings: Vec<String> = Vec::new();
+    let mut private_arg_strings: Vec<String> = Vec::new();
+    let mut defined_arg_names: HashSet<String> = HashSet::new();
+
+    for arg_part_pair in arg_section_pair.into_inner() {
+        match arg_part_pair.as_rule() {
+            Rule::public_arg_list => {
+                for arg_ident_pair in arg_part_pair
+                    .into_inner()
+                    .filter(|p| p.as_rule() == Rule::identifier)
+                {
+                    let arg_name = arg_ident_pair.as_str().to_string();
+                    if !defined_arg_names.insert(arg_name.clone()) {
+                        return Err(ProcessorError::DuplicateWildcard {
+                            name: arg_name,
+                            span: Some(get_span(&arg_ident_pair)),
+                        });
+                    }
+                    public_arg_strings.push(arg_name);
+                }
+            }
+            Rule::private_arg_list => {
+                for arg_ident_pair in arg_part_pair
+                    .into_inner()
+                    .filter(|p| p.as_rule() == Rule::identifier)
+                {
+                    let arg_name = arg_ident_pair.as_str().to_string();
+                    if !defined_arg_names.insert(arg_name.clone()) {
+                        return Err(ProcessorError::DuplicateWildcard {
+                            name: arg_name,
+                            span: Some(get_span(&arg_ident_pair)),
+                        });
+                    }
+                    private_arg_strings.push(arg_name);
+                }
+            }
+            Rule::private_kw | Rule::COMMENT | Rule::WHITESPACE => {}
+            _ if arg_part_pair.as_str() == "," => {}
+            _ => {
```

Comment:
The `arg_section` rule by grammar definition doesn't have more cases, so this should be `unreachable`.